### PR TITLE
add prediction diff for nonsymmetric trees

### PR DESCRIPTION
**Pull request by felixandrer**
Before submitting a pull request, please do the following steps:

1. Read [instructions for contributors](https://catboost.ai/docs/concepts/development-and-contributions.html).
2. Run `ya make` in catboost folder to make sure the code builds.
3. Add tests that test your change.
4. Run tests using `ya make -t -A` command.
5. If you haven't already, complete the [CLA](https://yandex.ru/legal/cla/).

### CODE REVIEW

File: catboost/libs/fstr/compare_documents.cpp [hunk 1]
```diff
@@ -45,8 +45,83 @@ static void CalcIndicatorCoefficients(
...
+        const TVector<double>& featureValues,
+        TArrayRef<ui32> treeLeafIdxes,
+        TVector<TVector<double>>* floatFeatureImpact
+) {
+    floatFeatureImpact->resize(model.ModelTrees->GetFloatFeatures().size());
+    for (ui32 featureIdx = 0; featureIdx < floatFeatureImpact->size(); ++featureIdx) {
+        (*floatFeatureImpact)[featureIdx].resize(model.ModelTrees->GetFloatFeatures()[featureIdx].Borders.size() + 1);
+    }
+
+    const auto treeSplits = model.ModelTrees->GetTreeSplits();
```
Comment (ek-ak) [inline catboost/libs/fstr/compare_documents.cpp:81]: const auto& treeSplits

File: catboost/libs/fstr/compare_documents.cpp [hunk 2]
```diff
@@ -45,8 +45,83 @@ static void CalcIndicatorCoefficients(
...
+    const auto stepNodes = model.ModelTrees->GetNonSymmetricStepNodes();
+
+    for (size_t treeId = 0; treeId < model.ModelTrees->GetTreeCount(); ++treeId) {
+        size_t leafOffset = model.ModelTrees->GetFirstLeafOffsets()[treeId];
+        double leafValue = model.ModelTrees->GetLeafValues()[leafOffset + treeLeafIdxes[treeId]];
+        size_t splitIdx = model.ModelTrees->GetTreeStartOffsets()[treeId];
+        int nextSplitStep = 0;
+        do {
+            auto split = model.ModelTrees->GetBinFeatures()[treeSplits[splitIdx]];
```
Comment (ek-ak) [inline catboost/libs/fstr/compare_documents.cpp:90]: const auto& split

File: catboost/libs/fstr/compare_documents.cpp [hunk 3]
```diff
@@ -88,17 +164,21 @@ TVector<TVector<double>> GetPredictionDiff(
...
+    CB_ENSURE(model.GetNumCatFeatures() == 0, "Model with categorical features are not supported");
 
     TVector<ui32> leafIdxes = CalcLeafIndexesMulti(model, dataProvider.ObjectsData, 0, 0);
 
     const auto *const rawObjectsData = dynamic_cast<const TRawObjectsDataProvider*>(dataProvider.ObjectsData.Get());
 
     const auto& binSplits = model.ModelTrees->GetBinFeatures();
 
+    TVector<TVector<double>> floatFeatureValues(dataProvider.GetObjectCount());
+    for (size_t idx=0; idx < dataProvider.GetObjectCount(); ++idx) {
```
Comment (ek-ak) [inline catboost/libs/fstr/compare_documents.cpp:178]: whitespaces around =

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@felixandrer, please agree to the CLA (https://github.com/catboost/catboost/blob/master/CONTRIBUTING.md#provide-contributions)

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request created: [1213991](https://a.yandex-team.ru/review/1213991)

**Comment by felixandrer** [comment]
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en.

**Comment by arcadia-devtools** [comment]
Check [1213991](https://a.yandex-team.ru/review/1213991) failed: 33 test(s) broken

**Comment by ek-ak** [comment]
@arcadia-devtools Ship it!

**Comment by Evgueni-Petrov-aka-espetrov** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@Evgueni-Petrov-aka-espetrov, internal review request updated: [1213991](https://a.yandex-team.ru/review/1213991)

**Comment by arcadia-devtools** [comment]
Check [1213991](https://a.yandex-team.ru/review/1213991) passed

**Pull request merged**